cluster/etcd_service_discovery.go

```diff
@@ -405,6 +417,76 @@ func (sd *etcdServiceDiscovery) printServers() {
 	}
 }
 
+// Struct that encapsulates a parallel/concurrent etcd get
+// it spawns goroutines and receives work requests through a channel
+type parallelGetterWork struct {
+	serverType string
+	serverID   string
+}
+
+type parallelGetter struct {
+	cli         *clientv3.Client
+	numWorkers  int
+	wg          *sync.WaitGroup
+	resultMutex sync.Mutex
+	result      *[]*Server
+	workChan    chan parallelGetterWork
+}
+
+func newParallelGetter(cli *clientv3.Client, numWorkers int) parallelGetter {
+	if numWorkers <= 0 {
+		numWorkers = 10
+	}
+	p := parallelGetter{
+		cli:        cli,
+		numWorkers: numWorkers,
+		workChan:   make(chan parallelGetterWork),
+		wg:         new(sync.WaitGroup),
+		result:     new([]*Server),
+	}
+	p.start()
+	return p
+}
+
+func (p *parallelGetter) start() {
+	for i := 0; i < p.numWorkers; i++ {
+		go func() {
+			for work := range p.workChan {
+				logger.Log.Debugf("loading info from missing server: %s/%s", work.serverType, work.serverID)
+
+				sv, err := getServerFromEtcd(p.cli, work.serverType, work.serverID)
+				if err != nil {
+					logger.Log.Errorf("error getting server from etcd: %s, error: %s", work.serverID, err.Error())
+					println("calling done")
+					p.wg.Done()
+					return
```

Comment:
do you really want to kill the goroutine in this case or just continue?

Reply:
good point, it should be a continue